rfcs/0028-nix-release-model.md

```diff
@@ -0,0 +1,127 @@
+---
+feature: nix-release-model
+start-date: 2018-03-28
+author: Shea Levy <shea@shealevy.com>
+co-authors: (find a buddy later to help our with the RFC)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+A model of when, what, and how to make new releases of Nix.
+
+# Motivation
+[motivation]: #motivation
+
+We want to set clear expectations for users, provide guidance for
+developers, and provide both groups the benefits of easy availability
+of new improvements.
+
+# Detailed design
+[design]: #detailed-design
+
+The proposed process contains a number of interrelated components.
+I've numbered them for ease of reference and discussion, not to
+indicate any meaningful ordering. Any aspect of this can be suspended
+at the security team's discretion when handling a vulnerability. Items
+marked with a * need technical work to implement.
+
+1. Keep `master` as close to release readiness as possible. This
+   includes testing, documentation, changelog entries, etc. being
+   ready *before* merge.
+2. *Use code review and automated checks, enforced through branch
+    protection, to ensure point 1 (and generally improve quality and
+    increase knowledge spread).
+3. Follow semver. In light of 1, this includes ensuring any necessary
+   semver bumps happen before a relevant change is merged.
+4. Any community member may request a new release by opening an issue
+   on GitHub. If any Nix core team member (who isn't the requestor)
+   thinks a release is warranted, they can assign themselves the
+   ticket and assume release manager responsibilities for the release.
+5. There will be at least one mainline release before each `NixOS`
+   stable release.
+6. The release manager will branch off a `#-stabilization` branch
+   (e.g. `2.0.1-stabilization`) in advance of any release.
+7. Changes aimed at getting the release ready (fixing bugs, improving
+   docs, etc.) should target the `#-stabilization` branch and then
+   be merged forward into `master`. No commit should exist on the
+   `#-stabilization` branch without very soon thereafter being merged
+   into `master`.
```

Comment:
Do you imply that if `master` drops some code that is still present in the stabilisation branch, the release process has to be restarted?

Reply:
Hmm... The idea of stabilization is that it should be fairly quick. If it's taking long enough that master wants to diverge this much, we've gone wrong IMO. Can you give an example of where what you're suggesting might happen?

Reply:
Well, there is not much that I can predict getting dropped… But, say, if derivation-files get dropped completely (which would be a pity for me but seems likely), I wouldn't be surprised if it happens soon after starting a release stabilisation branch. If by any chance it so happens that a new feature would be simpler to implement without derivation-files, there can be an incentive to drop in master before release is formally announced.

Reply:
So you're imagining a) we branch off a stabilization branch b) we remove derivation files on master*, c) we need a fix on stabilization that involves derivation files? I think that kind of thing is unlikely enough that we can just tell people to wait for stabilization... Maybe we should revisit if/when this arises.

*: IMO if we do remove derivation files it should still be possible to serialize derivations somehow, just maybe not by default.

Reply:
As written, (c) is guaranteed to happen: it will be the update to the documentation that replaces «is deprecated and will be dropped soon» with «will be dropped in the next release».

Reply:
Hmm... OK I'll add an exclusion for this, but I didn't intend for this to be *quite* so legalistic :grinning:

Reply:
Well, I am trying to balance nitpickiness and specific-scenario-realism. I think an exclusion specifically for documenting dropped features is not a good idea per se.

Maybe the answer depends on the answer to my global question: before-large-changes releases have larger risk of immediate divergence (whatever the subject matter is), after-large-changes releases have larger risk of missed consequences.